Bump minimum python version to 3.7

maflcko opened the pull request:
While there is nothing that requires a bump, it may require less maintenance to drop python3.6 support. Python3.7 is available through the package manager on all currently supported operating systems.

kristapsk commented:
Concept ACK, Python 3.6 had EOL 23 Dec 2021. https://endoflife.date/python

theStack commented on doc/dependencies.md:
```diff
@@ -10,7 +10,7 @@ You can find installation instructions in the `build-*.md` file for your platfor
 | [Clang](https://clang.llvm.org) | [8.0](https://github.com/bitcoin/bitcoin/pull/24164) |
 | [GCC](https://gcc.gnu.org) | [8.1](https://github.com/bitcoin/bitcoin/pull/23060) |
-| [Python](https://www.python.org) (tests) | [3.6](https://github.com/bitcoin/bitcoin/pull/19504) |
+| [Python](https://www.python.org) (tests) | [3.7](https://github.com/bitcoin/bitcoin/pull/19504) |
```
```suggestion
| [Python](https://www.python.org) (tests) | [3.7](https://github.com/bitcoin/bitcoin/pull/26226) |
```

MarcoFalke replied:
thx, done

theStack reviewed (commented):
Concept ACK

jarolrod reviewed (commented):
Concept ACK modulo https://github.com/bitcoin/bitcoin/pull/26226#discussion_r985309527

It would be useful to include a build system release note when dependencies like this are updated, rather than leaving it hidden among the change-log sea; but that's aside this pr.

fanquake commented:
Concept ACK for 25.x

MarcoFalke commented:
> Concept ACK, Python 3.6 had EOL 23 Dec 2021. https://endoflife.date/python

EOL of python shouldn't affect us, as we only use it to run the tests. Though, operating systems that ship with it going EOL would affect us.

fjahr commented:
I don't get the argument to do this now when nothing has changed since #24017...

Sjors commented:
No strong opinions. Paging @luke-jr.

MarcoFalke commented:
> I don't get the argument to do this now when nothing has changed since https://github.com/bitcoin/bitcoin/pull/24017...

Right, there is no rush, so I guess I better close this.

It just has to be done at some point in the future and I thought having a working patch couldn't hurt.

luke-jr commented:
>No strong opinions. Paging @luke-jr.

Thanks. I did take a look at this the other day. My notes are that 3.6 was for RHEL, but it appears RHEL has since moved on.

Personally, 3.6 is slightly more convenient to me since I have no easy way to get 3.7 anymore for testing (nor 3.6, but it's already installed) - Gentoo's oldest is 3.8. But I can adapt if need be.

jamesob commented on test/lint/lint-assertions.py:
```diff
@@ -12,7 +12,7 @@
 def git_grep(params: [], error_msg: ""):
     try:
-        output = subprocess.check_output(["git", "grep", *params], universal_newlines=True, encoding="utf8")
+        output = subprocess.check_output(["git", "grep", *params], text=True, encoding="utf8")
```
I think `encoding` is now unnecessary to specify in this line (and many others below) because `text` ensures stdout is a string and strings in Python 3 are by default unicode.

MarcoFalke replied:
Heh, no. This only changed in Python 3.7, see https://docs.python.org/3/library/os.html#utf8-mode

I guess `text`/`universal_newlines` is not needed when specifying `encoding`, so might as well remove them completely.

luke-jr replied:
Why wouldn't they be needed? Windows still has non-standard newlines.

jamesob reviewed (commented):
Concept ACK

MarcoFalke commented:
Closing for now until there is a more pressing reason

jamesob commented:
Can we reopen this? 
- There are some nice features in Python 3.7+ (like typed namedtuple features, `dataclasses`),
- Python 3.6 is EOL, and
- platforms like Debian stable ship with Python 3.9.

kristapsk commented:
> There are some nice features in Python 3.7+ (like typed namedtuple features, `dataclasses`),

Agree this could be reason to reopen and do this, if we see benefits from these features.

jamesob commented:
I think we should do anything we can to encourage people to add to and improve the functional test suite, and saddling test writers with an out of date version of python that makes it difficult to write modern idiomatic code because we're too lazy to review a test-only PR (?) seems like the opposite direction.

hebasto commented:
Concept ACK.

MarcoFalke commented:
Rebased, but no thoughts/opinion whether this should be done, or when a good time might be.

jamesob commented:
ACK

Why wait? What's the risk? 3.7 unlocks some nice features: https://docs.python.org/3.7/whatsnew/3.7.html

Notable are dataclasses, some `typing` improvements, and that dicts now preserve order based on insertion.

I can't really think about who would be fretting about minimimum Python version... if it's too contemporary for your local host, run the tests in a container.

fjahr commented:
Concept ACK

MarcoFalke commented:
> What's the risk?

Mostly users on Ubuntu Bionic or Centos 8 yelling that their default python3 can't run the tests

1440000bytes reviewed (approved):
ACK https://github.com/bitcoin/bitcoin/pull/26226/commits/f7459c8582dfd19a84aa8052d59f39babbe9d996

hebasto commented:
> Mostly users on Ubuntu Bionic or Centos 8 yelling that their default python3 can't run the tests

FWIW, the current minimum 3.6.15 is greater than Bionic's python version which is [3.6.7](https://packages.ubuntu.com/bionic-updates/python3).

MarcoFalke commented:
> FWIW, the current minimum version 3.6.15 is greater than Bionic's python version which is [3.6.7](https://packages.ubuntu.com/bionic-updates/python3).

3.6 is still the minimum version. And we check that on master by running the tests on Bionic.

achow101 commented:
> Mostly users on Ubuntu Bionic or Centos 8 yelling that their default python3 can't run the tests

Both of those also EOL this year.

jamesob commented:
I hereby volunteer to provide the code necessary to generate a container sufficient to run the functional tests for anyone both (i) yelling and (ii) running an ancient version of Python. :)

MarcoFalke commented:
Right, Ubuntu Bionic will be EOL before the next release anyway. Centos Stream 8 runs until  May 31st, 2024 , but they can install `python38`, I guess :man_shrugging:

MarcoFalke commented:
(Force pushed minimum to 3.7.16 and "maximum" to 3.12)

Commit:
scripted-diff: Use new python 3.7 keywords

-BEGIN VERIFY SCRIPT-
 sed -i 's/universal_newlines/text/g' $(git grep -l universal_newlines)
-END VERIFY SCRIPT-

Commit:
Bump minimum python version to 3.7

Commit:
Revert "contrib: Fix capture_output in getcoins.py"

This reverts commit be59bd17ec753af7cc763474f2432d12bfc88c2f
because the changes are no longer needed.

achow101 commented:
Since #26716 changes our CI to use pyenv's python, I think the changes to the container and installed packages in fa33242a1ab340aad177021f6f12073081615f86 end up being irrelevant?

MarcoFalke commented:
The ci/lint system is separate from the ci/test system, so you should be able to get a test failure by reverting the image/packages in ci/test

achow101 commented:
ACK fa1dbe04b2580a7d426be1729a3c78c61964075e

hebasto commented on ci/test/04_install.sh:
```diff
@@ -92,6 +92,9 @@ elif [ "$CI_USE_APT_INSTALL" != "no" ]; then
     CI_EXEC_ROOT add-apt-repository ppa:hadret/bpfcc
   fi
+  if [[ -n "${APPEND_APT_SOURCES_LIST}" ]]; then
+    CI_EXEC echo "${APPEND_APT_SOURCES_LIST}" >> /etc/apt/sources.list
```
fa33242a1ab340aad177021f6f12073081615f86:
```suggestion
    CI_EXEC_ROOT echo "${APPEND_APT_SOURCES_LIST}" >> /etc/apt/sources.list
```

hebasto reviewed (approved):
ACK fa8fe5b69669b4d86e0d0970d68502abee8785f3

jamesob commented:
ACK https://github.com/bitcoin/bitcoin/pull/26226/commits/fa8fe5b69669b4d86e0d0970d68502abee8785f3

Looks ready to go to me!

jamesob commented:
As a follow-up I'm betting we can drop all the changes that use `pyvm` to build Python from source, since Python 3.7 comes with buster by default?

MarcoFalke commented:
No, we also need git2.34 or later, which is only shipped in Ubuntu Jammy

willcl-ark commented:
Post merge ACK.

FWIW my version of `pyenv` did not contain 3.7.16 in the list of versions available to install. Easily fixed though by running `pyenv update` to update both `pyenv` itself, and also the list of available versions for install.
